File: tests/cpu/seats/danube-rows.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt } from '../helpers';
import { emptySeat, warBanned, diploVisibility, visibilityCS } from '../../../cpu/core/seats';
import { getModifiers, progressAhead } from '../../../cpu/core/effects';
import { tourismFavorOf } from '../../../cpu/core/effects';
import { tradeCapacity, intlRouteTerrainYields, stampTradingPost } from '../../../cpu/core/trade';
import { declareWarOnCityState, setMet, placeCityStateAt } from '../../../cpu/core/cityStates';
import { CIV_LEADERS } from '../../../cpu/data/seats';
import {
  WONDER_ERA_PROD_ROWS, WONDER_TOURISM_ROWS, RIVER_CROSS_PROD_ROWS, IMMEDIATE_POST_ROWS,
  DIPLO_VIS_ROWS, WAR_BAN_ROWS, WAR_BANS, TOURISM_FAVOR_ROWS, EMERGENCY_FAVOR_ROWS,
  GOLDEN_DEDICATION_ROWS, INTL_ROUTE_TERRAIN_ROWS, GOLDEN_ROUTE_CAPACITY_ROWS, PROGRESS_TRADE_ROWS,
} from '../../../cpu/data/civilizations';
import type { GameState } from '../../../cpu/core/types';

/**
 * THE WONDER, THE RIVER AND THE POST (CIV6, the install's TraitModifiers):
 * France's wonder band and tourism, Pearl of the Danube, Ortoo, Faces of
 * Peace, Strength in Unity, Sahel Merchants and the Grand Embassy.
 *
 * The GPU twin is tests/gpu/danube_rows_test.py.
 */
const PLAIN = -1;
const seatRow = (civ: string) => CIV_LEADERS.findIndex((l) => l.civ === civ);
const leaderRow = (leader: string) => CIV_LEADERS.findIndex((l) => l.leader === leader);

function sceneAs(row: number): GameState {
  const state = makeState(makeMap(16, 16, 'GRASSLAND'));
  state.seats.push(emptySeat(1));
  state.seats[0].civ = row;
  state.seats[1].civ = seatRow('AMERICA');
  return state;
}

describe('the wire', () => {
  it('carries every batch-twelve family, with the install\'s own magnitudes', () => {
    expect(WONDER_ERA_PROD_ROWS.length).toBe(1);
    expect(WONDER_TOURISM_ROWS.length).toBe(1);
    expect(RIVER_CROSS_PROD_ROWS.length).toBe(2);
    expect(IMMEDIATE_POST_ROWS.length).toBe(1);
    // the family grew with Catherine's flat level; a census pin is a count
    expect(DIPLO_VIS_ROWS.length).toBe(2);
    expect(WAR_BAN_ROWS.length).toBe(3);
    expect(TOURISM_FAVOR_ROWS.length).toBe(1);
    expect(EMERGENCY_FAVOR_ROWS.length).toBe(1);
    expect(GOLDEN_DEDICATION_ROWS.length).toBe(1);
    expect(INTL_ROUTE_TERRAIN_ROWS.length).toBe(1);
    expect(GOLDEN_ROUTE_CAPACITY_ROWS.length).toBe(1);
    expect(PROGRESS_TRADE_ROWS.length).toBe(1);
    // the band is INCLUSIVE at both ends, and the install's own era names
    expect(WONDER_ERA_PROD_ROWS[0].startEra).toBe('Medieval');
    expect(WONDER_ERA_PROD_ROWS[0].endEra).toBe('Industrial');
    // the visibility row ADDS a second step rather than replacing the first
    expect(DIPLO_VIS_ROWS.find((r) => r.civ === 'MONGOLIA')!.csPerLevel).toBe(3);
    for (const r of WAR_BAN_ROWS) expect(WAR_BANS.indexOf(r.ban)).toBeGreaterThanOrEqual(0);
  });
});

describe('Faces of Peace', () => {
  it('refuses Canada a surprise war, and refuses one declared ON Canada', () => {
    const canada = sceneAs(seatRow('CANADA'));
    const plain = sceneAs(PLAIN);
    // a SURPRISE war is the informal one
    expect(warBanned(canada, 0, 1, false)).toBe(true);
    expect(warBanned(canada, 0, 1, true)).toBe(false); // a formal war is fine
    expect(warBanned(plain, 0, 1, false)).toBe(false);
    // ...and nobody may surprise Canada either
    const other = sceneAs(PLAIN);
    other.seats[1].civ = seatRow('CANADA');
    expect(warBanned(other, 0, 1, false)).toBe(true);
    expect(warBanned(other, 0, 1, true)).toBe(false);
  });

  it('refuses Canada a war on a city-state, and refuses nobody else', () => {
    const attempt = (row: number) => {
      const state = sceneAs(row);
      settleAt(state, tileAtCoords(state.map, 4, 4).index, 0);
      // the bare scene seats no minor, so the ban would never be reached
      const cs = placeCityStateAt(state, 0, 'Testopolis', 'militaristic',
                                  tileAtCoords(state.map, 12, 12).index);
      setMet(cs, 0);
      return declareWarOnCityState(state, cs.id, 0);
    };
    const canada = attempt(seatRow('CANADA'));
    expect(canada.ok).toBe(false);
    expect(canada.reason).toContain('may not declare war on a city-state');
    expect(attempt(PLAIN).ok).toBe(true);
  });

  it('turns every hundred Tourism per turn into a Favor', () => {
    const canada = sceneAs(seatRow('CANADA'));
    const plain = sceneAs(PLAIN);
    expect(tourismFavorOf(canada, 0, 250)).toBe(2);
    expect(tourismFavorOf(canada, 0, 99)).toBe(0);
    expect(tourismFavorOf(plain, 0, 250)).toBe(0);
  });

  it('doubles the favor an emergency pays it', () => {
    expect(getModifiers(sceneAs(seatRow('CANADA')), 0).emergencyFavorPct).toBe(100);
    expect(getModifiers(sceneAs(PLAIN), 0).emergencyFavorPct).toBe(0);
  });
});

describe('Ortoo', () => {
  it('reads a trading post as an extra level of visibility', () => {
    const level = (row: number, post: boolean): number => {
      const state = sceneAs(row);
      settleAt(state, tileAtCoords(state.map, 4, 4).index, 0);
      const theirs = settleAt(state, tileAtCoords(state.map, 12, 12).index, 1);
      if (post) stampTradingPost(state.seats[0], theirs.centerIndex);
      return diploVisibility(state, 0, 1);
    };
    expect(level(seatRow('MONGOLIA'), false)).toBe(level(PLAIN, false));
    expect(level(seatRow('MONGOLIA'), true)).toBe(level(PLAIN, true) + 1);
  });

  it('doubles the strength a visibility advantage pays', () => {
    const cs = (row: number): number => {
      const state = sceneAs(row);
      settleAt(state, tileAtCoords(state.map, 4, 4).index, 0);
      settleAt(state, tileAtCoords(state.map, 12, 12).index, 1);
      // one level of advantage: the viewer holds the tech, the target does not
      state.seats[0].research.techs = ['PRINTING'];
      return visibilityCS(state, 0, 1);
    };
    const plain = cs(PLAIN);
    expect(plain).toBeGreaterThan(0);
    expect(cs(seatRow('MONGOLIA'))).toBe(plain * 2);
  });
});

describe('Sahel Merchants', () => {
  it('pays a Gold per flat Desert tile of the origin city, and none on hills', () => {
    const gold = (row: number, hills: boolean): number => {
      const state = sceneAs(row);
      const city = settleAt(state, tileAtCoords(state.map, 8, 8).index, 0);
      let n = 0;
      for (const t of state.map.tiles) {
        if (t.ownerSeat !== 0 || t.ownerCity !== city.id) continue;
        t.terrain = 'DESERT';
        t.elevation = hills ? 'HILLS' : 'FLAT';
        n += 1;
      }
      expect(n).toBeGreaterThan(0);
      return intlRouteTerrainYields(state, city, 0).gold;
    };
    expect(gold(leaderRow('MANSA_MUSA'), false)).toBeGreaterThan(0);
    expect(gold(leaderRow('MANSA_MUSA'), true)).toBe(0); // the install names FLAT desert
    expect(gold(PLAIN, false)).toBe(0);
  });

  it('adds a Trade Capacity per golden age entered', () => {
    const cap = (row: number, ages: number): number => {
      const state = sceneAs(row);
      settleAt(state, tileAtCoords(state.map, 8, 8).index, 0);
      state.seats[0].goldenAges = ages;
      return tradeCapacity(state, 0);
    };
    expect(cap(leaderRow('MANSA_MUSA'), 0)).toBe(cap(PLAIN, 0));
    expect(cap(leaderRow('MANSA_MUSA'), 2)).toBe(cap(PLAIN, 2) + 2);
  });
});

describe('the Grand Embassy', () => {
  it('counts only how far AHEAD the other seat is', () => {
    const state = sceneAs(leaderRow('PETER_GREAT'));
    state.seats[0].research.techs = ['MINING'];
    state.seats[1].research.techs = ['MINING', 'POTTERY', 'SAILING', 'WRITING'];
    expect(progressAhead(state, 0, 1, false)).toBe(3);
    expect(progressAhead(state, 1, 0, false)).toBe(0); // never negative
    expect(getModifiers(state, 0).progressTradePer).toBe(3);
    expect(getModifiers(sceneAs(PLAIN), 0).progressTradePer).toBe(0);
  });
});

File: tests/gpu/danube_rows_test.py
"""THE WONDER, THE RIVER AND THE POST — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/danube_rows_test.py

The TS twin is tests/cpu/seats/danube-rows.test.ts.

CIV6 (the install's TraitModifiers): France's wonder band and tourism, Pearl
of the Danube, Ortoo, Faces of Peace, Strength in Unity, Sahel Merchants and
the Grand Embassy.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

B0 = 0
RULES = json.loads((Path(__file__).resolve().parent.parent.parent
                    / "seeder" / "worlds" / "rules.json").read_text())
TECHS = [t["id"] for t in RULES["techs"]]
TERRAINS = RULES["terrains"] if isinstance(RULES.get("terrains"), list) else None


def play(sim, row: int, name):
    """Seat one roster row, or clear it (`name=None`) for the BASELINE."""
    if name is None:
        sim.row_civ[0, row] = -1
        sim.row_leader[0, row] = -1
    else:
        ci = sim._civ_ids.index(name)
        sim.row_civ[0, row] = ci
        sim.row_leader[0, row] = sim._pair_civ.index(ci)
    sim._eff_version += 1
    sim._gen_ver += 1
    sim._bldg_version += 1
    if sim._gov_pol_cache is not None:
        sim._gov_pol_cache.clear()


def lead(sim, row: int, civ: str, leader: str) -> None:
    play(sim, row, civ)
    sim.row_leader[0, row] = sim._leader_idx(leader)
    sim._eff_version += 1


def fresh(rules, path) -> BatchSim:
    sim = BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64)
    for r, name in enumerate(("ROME", "EGYPT", "NORWAY")):
        play(sim, r, name)
    return settle_all(sim)


def seat(sim, row: int, name, leader=None):
    if leader is None:
        play(sim, row, name)
    else:
        lead(sim, row, name, leader)


# ---------------------------------------------------------------------------


def test_wire(rules, path) -> None:
    sim = fresh(rules, path)
    assert len(sim._wonder_era_prod_rows) == 1 and len(sim._wonder_tourism_rows) == 1
    assert len(sim._river_cross_prod_rows) == 2 and len(sim._immediate_post_rows) == 1
    # the visibility family grew with Catherine's flat level
    assert len(sim._diplo_vis_rows) == 2 and len(sim._war_ban_rows) == 3
    assert len(sim._tourism_favor_rows) == 1 and len(sim._emergency_favor_rows) == 1
    assert len(sim._golden_dedication_rows) == 1
    assert len(sim._intl_route_terrain_rows) == 1
    assert len(sim._golden_route_capacity_rows) == 1 and len(sim._progress_trade_rows) == 1
    # the band is inclusive at both ends, on the engine's own era indices
    _fc, _fl, _fs, _fe, _fp = sim._wonder_era_prod_rows[0]
    assert 0 <= _fs < _fe and _fp == 20
    # the visibility row ADDS a second step rather than replacing the first
    assert max(r[3] for r in sim._diplo_vis_rows) == sim._vis_cs_per_level
    print("  1 wire OK — 12 families, and the visibility step doubles")


def test_faces_of_peace(rules, path) -> None:
    """Canada declares no surprise war, and takes none."""
    sim = fresh(rules, path)
    if sim.n_majors < 2:
        print("  2 Faces of Peace SKIPPED — one major only")
        return

    def declared(name_a, name_b) -> bool:
        s2 = fresh(rules, path)
        seat(s2, 0, name_a)
        seat(s2, 1, name_b)
        s2.war[B0, 0, 1] = False
        s2.war[B0, 1, 0] = False
        s2.treaty_turns[B0, 0, 1] = 0
        s2.seat_denounced[B0, 0, 1] = 0          # no casus belli -> a SURPRISE war
        s2._declare_war_major(0, 1, torch.ones(s2.B, dtype=torch.bool))
        return bool(s2.war[B0, 0, 1])

    assert declared(None, None), "a plain seat could not declare"
    assert not declared("CANADA", None), "Canada declared a surprise war"
    assert not declared(None, "CANADA"), "a surprise war was declared ON Canada"
    print("  2 Faces of Peace OK — no surprise war either way, and others declare freely")


def test_tourism_favor(rules, path) -> None:
    """A Favor per hundred Tourism a turn."""
    def favor(name, rate: int) -> int:
        s2 = fresh(rules, path)
        seat(s2, 0, name)
        s2.civ_tour_rate[B0, 0] = rate
        s2._eff_version += 1
        return int(s2._tourism_favor_of(0)[B0])

    assert favor("CANADA", 250) == 2, "250 tourism did not pay 2"
    assert favor("CANADA", 99) == 0, "under a hundred paid"
    assert favor(None, 250) == 0, "a plain seat took the favor"
    print("  3 the tourism favor OK — 2 at 250, 0 at 99, none for a plain seat")


def test_ortoo(rules, path) -> None:
    """A trading post is a level of sight, and the step doubles."""
    sim = fresh(rules, path)
    if sim.n_majors < 2:
        print("  4 Ortoo SKIPPED — one major only")
        return

    def level(name, post: bool) -> int:
        s2 = fresh(rules, path)
        seat(s2, 0, name)
        s2.trading_post[B0] = False
        if post:
            _ctr = int(s2.city_center[B0, 1, 0])
            assert _ctr >= 0, "the target seats no city"
            s2.trading_post[B0, 0, _ctr] = True
        s2._eff_version += 1
        return int(s2._diplo_vis()[B0, 0, 1])

    assert level("MONGOLIA", False) == level(None, False), "a postless Mongol saw more"
    assert level("MONGOLIA", True) == level(None, True) + 1, "the post paid no level"

    def cs(name) -> int:
        s2 = fresh(rules, path)
        seat(s2, 0, name)
        s2.civ_techs[B0, :, :] = False
        if s2._vis_tech >= 0:
            s2.civ_techs[B0, 0, s2._vis_tech] = True   # one level of advantage
        s2._eff_version += 1
        return int(s2._vis_cs(torch.zeros(s2.B, dtype=torch.long),
                              torch.ones(s2.B, dtype=torch.long))[B0])

    plain = cs(None)
    assert plain > 0, "the baseline had no visibility advantage, so the scene proves nothing"
    assert cs("MONGOLIA") == plain * 2, f"Mongolia read {cs('MONGOLIA')} against {plain}"
    print(f"  4 Ortoo OK — the post is a level, and {plain} doubles to {plain * 2}")


def test_sahel_merchants(rules, path) -> None:
    """A Trade Capacity per golden age entered."""
    def cap(name, leader, ages: int) -> int:
        s2 = fresh(rules, path)
        seat(s2, 0, name, leader)
        s2.golden_ages[B0, 0] = ages
        s2._eff_version += 1
        return int(s2._trade_capacity(0)[B0])

    assert cap("MALI", "MANSA_MUSA", 0) == cap(None, None, 0), "a golden-less seat gained capacity"
    assert cap("MALI", "MANSA_MUSA", 2) == cap(None, None, 2) + 2, "two golden ages paid nothing"
    print("  5 Sahel Merchants OK — one Trade Capacity per golden age entered")


def test_strength_in_unity(rules, path) -> None:
    """A GOLDEN age still pays the normal era score, for this row alone."""
    def score(name, age: int) -> float:
        s2 = fresh(rules, path)
        seat(s2, 0, name)
        if s2.ded_picks.shape[2] == 0:
            return -1.0
        # COMMIT one: the fixture picks none this early, and a skipped scene
        # proves nothing about the guard this row reaches past
        kind = next((k for k in range(len(s2._ded_event_score))
                     if float(s2._ded_event_score[k]) > 0), -1)
        if kind < 0:
            return -1.0
        s2.ded_picks[B0, 0, 0] = kind
        s2.civ_age[B0, 0] = age
        before = float(s2.era_score[B0, 0])
        s2._dedication_event(0, kind, torch.ones(s2.B, dtype=torch.long))
        return float(s2.era_score[B0, 0]) - before

    plain_norm = score(None, 1)
    if plain_norm < 0:
        print("  6 Strength in Unity SKIPPED — this scene commits no dedication")
        return
    assert plain_norm > 0, "a normal age paid no era score"
    assert score(None, 2) == 0.0, "a plain seat scored in a golden age"
    assert score("GEORGIA", 2) == plain_norm, "Georgia did not keep the normal bonus"
    print("  6 Strength in Unity OK — the golden age still pays the normal era score")


def test_wonder_tourism_is_per_game(rules, path) -> None:
    """The roster mask is [B] — ONE answer per game in the batch. A batch
    where a single game seats France must not pay every game: a collapsed
    `.any()` here doubled a neighbouring game's tourism, and only a batched
    serve shard could see it (single-seed runs stayed green)."""
    sim = BatchSim([load_fixture(path), load_fixture(path)], rules,
                   device="cpu", dtype=torch.float64)
    _c, _l, _p = sim._wonder_tourism_rows[0]
    row = 1
    # game 0 seats nobody on this row; game 1 seats the carrier
    sim.row_civ[0, row] = -1
    sim.row_leader[0, row] = -1
    sim.row_civ[1, row] = _c
    sim.row_leader[1, row] = sim._pair_civ.index(_c) if _c >= 0 else _l
    sim._eff_version += 1
    got = sim._wonder_tourism_pct(row).tolist()
    assert got == [0, _p], f"per-game percentage read {got}, expected [0, {_p}]"
    # ...and the row is per ROW too: another row of the same games takes none
    other = 0 if row else 1
    assert sim._wonder_tourism_pct(other).tolist() == [0, 0], (
        "a row that does not seat the carrier was paid")
    print("  7 wonder tourism OK — per game and per row, never batch-wide")

def main() -> int:
    rules = load_rules()
    path = fixture_paths()[0]
    test_wire(rules, path)
    test_faces_of_peace(rules, path)
    test_tourism_favor(rules, path)
    test_ortoo(rules, path)
    test_sahel_merchants(rules, path)
    test_strength_in_unity(rules, path)
    test_wonder_tourism_is_per_game(rules, path)
    print("BATTERY OK danube_rows")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
